Load tensors directly from pickle archive

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23281 Load tensors directly from pickle archive**

This allows tensors to be loaded directly from the pickle archive,
using the same encoding as the eager serializer. The one difference
is that the tensor data is still stored in a zip file rather
than appended to the archive, which still allows for aligning the tensor data.

With this change, we no longer rely on the json file for any information, so this PR also stops writing the json file and loads modules directly from the pickle.

New layout is:

* code/ - all code goes in these folders with python names
* constants{.pkl,/} - pickled code constants, these appear in the code so they cannot go in data, otherwise it would create a circular dependency.
* data{.pkl,/} - pickled data, with tensors in the data/ folder.

Differential Revision: [D16452815](https://our.internmc.facebook.com/intern/diff/D16452815)

zdevito committed:
[WIP] Load tensors directly from pickle archive

This allows tensors to be loaded directly from the pickle archive,
using the same encoding as the eager serializer. The one difference
is that the tensor data is still stored in a zip file rather
than appended to the archive.

zdevito committed:
Update on "[WIP] Load tensors directly from pickle archive"

[WIP] Load tensors directly from pickle archive

This allows tensors to be loaded directly from the pickle archive,
using the same encoding as the eager serializer. The one difference
is that the tensor data is still stored in a zip file rather
than appended to the archive.

gh-metadata: pytorch pytorch 23281 gh/zdevito/78/head

dzhulgakov commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -723,6 +729,48 @@ OpCode Unpickler::readInstruction() {
+          result = autograd::make_variable(result, requires_grad);
+          stack_.push_back(std::move(result));
+        });
+      } else if (module_name == "collections" && class_name == "OrderedDict") {
```
this is a bit scary. Is there are stricter condition we can enforce here (e.g. that it's indeed in Tensor deserialization only)

dzhulgakov commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -723,6 +729,48 @@ OpCode Unpickler::readInstruction() {
           }
         });
+      } else if (
+          module_name == "torch._utils" && class_name == "_rebuild_tensor_v2") {
```
would be nice to extract to a subfunction and explicitly document in both python and C++ that those functions should match semantically

dzhulgakov commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -723,6 +729,48 @@ OpCode Unpickler::readInstruction() {
               AT_ERROR("Unknown pickler class id");
           }
         });
+      } else if (
```
we should handle `_rebuild_qtensor` too I guess. cc @jerryzh168

zdevito committed:
Update on "[WIP] Load tensors directly from pickle archive"

Load tensors directly from pickle archive

This allows tensors to be loaded directly from the pickle archive,
using the same encoding as the eager serializer. The one difference
is that the tensor data is still stored in a zip file rather
than appended to the archive.

gh-metadata: pytorch pytorch 23281 gh/zdevito/78/head

zdevito committed:
Update on "Load tensors directly from pickle archive"

Load tensors directly from pickle archive

This allows tensors to be loaded directly from the pickle archive,
using the same encoding as the eager serializer. The one difference
is that the tensor data is still stored in a zip file rather
than appended to the archive.

gh-metadata: pytorch pytorch 23281 gh/zdevito/78/head

suo commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -748,6 +750,63 @@ OpCode Unpickler::readInstruction() {
+          module_name == "torch._utils" &&
+          (class_name == "_rebuild_tensor_v2" ||
+           class_name == "_rebuild_qtensor")) {
+        bool quantized = class_name == "_rebuild_qtensor";
```
ugh, I don't like that quantized tensors are somehow special case for the serialization format

zdevito replied:
I agree, but this is how it works in eager serialization, so we are stuck with it if we want torch.save from torchscript to work with torch.load

suo commented on torch/csrc/jit/pickle.h:
```diff
@@ -58,10 +58,9 @@ TORCH_API void pickle(
 ///
 /// See `torch::pickle` for details.
 TORCH_API IValue unpickle(
```
In general we should think a bit about how we want the Pickle C++ API to look at the end of this. @driazati, any thoughts? At least I think we should get rid of `tensor_table` if possible

driazati replied:
Any use of the `tensor_table` would be covered by pickle's [`persistent_id/load` functions](https://docs.python.org/2/library/pickle.html#pickling-and-unpickling-external-objects) (which we could implement if we need to in the future). This is already eager's serialization works anyways, so it's fine to get rid of the `tensor_table`

driazati commented on torch/csrc/jit/export.cpp:
```diff
@@ -618,34 +618,75 @@ class ScriptModuleSerializer {
+
+  void writeArchive(const std::string& archive_name, const IValue& value) {
+    std::vector<char> data;
+    Pickler data_pickle(
```
It looks like `jit::pickle(...)` would be fine here instead of the manual use of the `Pickler`

zdevito replied:
Right after this it uses `data_pickle.tensorData` which is not exposed by those functions.

driazati commented on torch/csrc/jit/import.cpp:
```diff
@@ -125,62 +125,84 @@ script::Module ScriptModuleDeserializer::deserialize(
+    ss << archive_name << "/" << name;
+    return std::get<0>(reader_->getRecord(ss.str()));
+  };
+  Unpickler unpickler(
```
Same here, this can be `jit::unpickle`

zdevito replied:
user-exposed pickle functions do not take the `read_record` or `device_` arguments needed here to load the tensors.

zdevito commented on torch/csrc/jit/pickler.h:
```diff
@@ -233,16 +237,29 @@ class Unpickler {
 
  public:
+  // tensors inside the pickle are references to the tensor_table
   Unpickler(
```
It is possible that RPC will be using this version, and completely handling tensor serialization on its own. Not sure though.

zdevito committed:
Update on "Load tensors directly from pickle archive"

Load tensors directly from pickle archive

This allows tensors to be loaded directly from the pickle archive,
using the same encoding as the eager serializer. The one difference
is that the tensor data is still stored in a zip file rather
than appended to the archive.

gh-metadata: pytorch pytorch 23281 gh/zdevito/78/head
